flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogStorage.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.changelog.fs;
+
+import org.apache.flink.annotation.Experimental;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.state.KeyGroupRange;
+import org.apache.flink.runtime.state.changelog.ChangelogStateHandleStreamImpl;
+import org.apache.flink.runtime.state.changelog.StateChangelogHandleReader;
+import org.apache.flink.runtime.state.changelog.StateChangelogHandleStreamHandleReader;
+import org.apache.flink.runtime.state.changelog.StateChangelogStorage;
+import org.apache.flink.util.ExceptionUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.flink.changelog.fs.FsStateChangelogOptions.PREEMPTIVE_PERSIST_THRESHOLD;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** Filesystem-based implementation of {@link StateChangelogStorage}. */
+@Experimental
+public class FsStateChangelogStorage
+        implements StateChangelogStorage<ChangelogStateHandleStreamImpl>, Serializable {
+    private static final Logger LOG = LoggerFactory.getLogger(FsStateChangelogStorage.class);
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * The log id is only needed on write to separate changes from different backends (i.e.
+     * operators) in the resulting file.
+     */
+    private transient AtomicInteger logIdGenerator = new AtomicInteger(0);
+
+    private transient volatile StateChangeUploader uploader;
+    private transient volatile FsStateChangelogCleaner cleaner;
+    private volatile ReadableConfig config; // todo: make final after FLINK-21804
```

Comment:
Can you explain what is happening here? Why do we need 3 volatiles, `AtomicInteger` and `synchronized (this)`?

Reply:
All this was supposed to go away after FLINK-21804 (as I mentioned in the comment - but I should definitely have elaborate more).
FLINK-21804 is now merged, so I'll rework this part after rebasing the PR.

Reply:
After dropping 25d1aacca8af5b70f97ee3843c491ac0ffa7d6bd there is only `AtomicLong` here.
It is needed because this storage can be used by different Tasks.

Reply:
Thanks for the explanation. It makes sense to me now.